polkadot/xcm/pallet-derivatives/src/lib.rs

```diff
@@ -0,0 +1,403 @@
+	fn try_register_derivative(
+		original: &OriginalOf<T, I>,
+		derivative: &DerivativeOf<T, I>,
+	) -> DispatchResult {
+		ensure!(
+			Self::original_to_derivative(original).is_none(),
+			Error::<T, I>::DerivativeAlreadyExists,
+		);
+
+		<OriginalToDerivative<T, I>>::insert(original, derivative);
+		<DerivativeToOriginal<T, I>>::insert(derivative, original);
+
+		Self::deposit_event(Event::<T, I>::DerivativeCreated { original: original.clone() });
+
+		Ok(())
+	}
+
+	fn try_deregister_derivative_of(original: &OriginalOf<T, I>) -> DispatchResult {
+		let derivative = <OriginalToDerivative<T, I>>::take(&original)
+			.ok_or(Error::<T, I>::NoDerivativeToDeregister)?;
+
+		<DerivativeToOriginal<T, I>>::remove(&derivative);
+		<DerivativeExtra<T, I>>::remove(&derivative);
+
+		Self::deposit_event(Event::<T, I>::DerivativeDestroyed { original: original.clone() });
+
+		Ok(())
+	}
+
+	fn get_derivative(original: &OriginalOf<T, I>) -> Result<DerivativeOf<T, I>, DispatchError> {
+		<OriginalToDerivative<T, I>>::get(original).ok_or(Error::<T, I>::DerivativeNotFound.into())
+	}
+
+	fn get_original(derivative: &DerivativeOf<T, I>) -> Result<OriginalOf<T, I>, DispatchError> {
+		<DerivativeToOriginal<T, I>>::get(derivative).ok_or(Error::<T, I>::OriginalNotFound.into())
+	}
+}
+
+impl<T: Config<I>, I: 'static> IterDerivativesRegistry<OriginalOf<T, I>, DerivativeOf<T, I>>
+	for Pallet<T, I>
+{
+	fn iter_originals() -> impl Iterator<Item = OriginalOf<T, I>> {
+		<OriginalToDerivative<T, I>>::iter_keys()
+	}
+
+	fn iter_derivatives() -> impl Iterator<Item = DerivativeOf<T, I>> {
+		<OriginalToDerivative<T, I>>::iter_values()
+	}
+
+	fn iter() -> impl Iterator<Item = (OriginalOf<T, I>, DerivativeOf<T, I>)> {
+		<OriginalToDerivative<T, I>>::iter()
+	}
+}
+
+impl<T: Config<I>, I: 'static> DerivativesExtra<DerivativeOf<T, I>, DerivativeExtraOf<T, I>>
+	for Pallet<T, I>
+{
+	fn get_derivative_extra(derivative: &DerivativeOf<T, I>) -> Option<DerivativeExtraOf<T, I>> {
+		<DerivativeExtra<T, I>>::get(derivative)
+	}
+
+	fn set_derivative_extra(
+		derivative: &DerivativeOf<T, I>,
+		extra: Option<DerivativeExtraOf<T, I>>,
+	) -> DispatchResult {
+		ensure!(
+			<DerivativeToOriginal<T, I>>::contains_key(derivative),
+			Error::<T, I>::DerivativeNotFound,
+		);
+
+		<DerivativeExtra<T, I>>::set(derivative, extra);
+
+		Ok(())
+	}
+}
+
+pub trait WeightInfo {
+	fn create_derivative() -> Weight;
+	fn destroy_derivative() -> Weight;
+}
+
+pub struct TestWeightInfo;
+impl WeightInfo for TestWeightInfo {
+	fn create_derivative() -> Weight {
+		Weight::from_parts(100_000_000, 0)
+	}
+
+	fn destroy_derivative() -> Weight {
+		Weight::from_parts(100_000_000, 0)
+	}
+}
+
+/// The `NoStoredMapping` adapter calls the `CreateOp` (which should take the `Original` value and
+/// return a `Derivative` one) and returns `None`, indicating that the mapping between the original
+/// and the derivative shouldn't be saved.
+pub struct NoStoredMapping<CreateOp>(PhantomData<CreateOp>);
```

Comment:
Why might you want to not store the mapping? Because the ids are the same? Can you add that to this doc comment?

Reply:
Yes, when IDs are the same, we don't need a mapping. In this scenario, the pallet only provides an API for derivative creation/destruction.

I'll highlight this in the doc comment